### ISSUE TITLE
Putting a question here to clear a doubt in sample thread pools implementation

### ISSUE THREAD

**Issue by rudraksh97** [arpitbbhayani/concurrency-in-depth#6]
Go version : 1.21.5

Doubt: seems like one distinct job is being done multiple times rather than 98 distinct jobs.

I modified the main function

```
func main() {
	pool := NewPool(7)
	for i := 0; i < 98; i++ {
		job := func() {
			time.Sleep(2 * time.Second)
			fmt.Printf("job: completed " + strconv.Itoa(i) + " " + time.Now().String() + "\n")
		}
		pool.AddJob(job)
	}
	pool.Wait()
}
```

output that I am getting has exact 98 entries but we see duplicate entries (i wise):
job: completed 7 2024-02-09 01:09:26.880134 +0530 IST m=+2.001490959
job: completed 7 2024-02-09 01:09:26.880078 +0530 IST m=+2.001435376
job: completed 7 2024-02-09 01:09:26.880265 +0530 IST m=+2.001622376
job: completed 7 2024-02-09 01:09:26.880277 +0530 IST m=+2.001633918
job: completed 7 2024-02-09 01:09:26.880283 +0530 IST m=+2.001640334
job: completed 7 2024-02-09 01:09:26.880104 +0530 IST m=+2.001461001
job: completed 7 2024-02-09 01:09:26.88054 +0530 IST m=+2.001896793
job: completed 14 2024-02-09 01:09:28.882469 +0530 IST m=+4.003833418
job: completed 14 2024-02-09 01:09:28.882456 +0530 IST m=+4.003821001
job: completed 14 2024-02-09 01:09:28.882552 +0530 IST m=+4.003917251
job: completed 14 2024-02-09 01:09:28.882556 +0530 IST m=+4.003921043
job: completed 14 2024-02-09 01:09:28.88256 +0530 IST m=+4.003925126
job: completed 14 2024-02-09 01:09:28.882564 +0530 IST m=+4.003928834
job: completed 14 2024-02-09 01:09:28.882571 +0530 IST m=+4.003935709
job: completed 21 2024-02-09 01:09:30.883693 +0530 IST m=+6.005065668
job: completed 22 2024-02-09 01:09:30.883839 +0530 IST m=+6.005211668
job: completed 23 2024-02-09 01:09:30.883849 +0530 IST m=+6.005222043
job: completed 24 2024-02-09 01:09:30.883864 +0530 IST m=+6.005236584
job: completed 25 2024-02-09 01:09:30.883871 +0530 IST m=+6.005244126
job: completed 26 2024-02-09 01:09:30.883882 +0530 IST m=+6.005254293
job: completed 21 2024-02-09 01:09:30.883666 +0530 IST m=+6.005039084
job: completed 28 2024-02-09 01:09:32.885989 +0530 IST m=+8.007369418
job: completed 29 2024-02-09 01:09:32.886359 +0530 IST m=+8.007739459
job: completed 30 2024-02-09 01:09:32.886377 +0530 IST m=+8.007757584
job: completed 31 2024-02-09 01:09:32.886385 +0530 IST m=+8.007765334
job: completed 32 2024-02-09 01:09:32.886422 +0530 IST m=+8.007802293
job: completed 33 2024-02-09 01:09:32.886428 +0530 IST m=+8.007809043
job: completed 34 2024-02-09 01:09:32.8865 +0530 IST m=+8.007880209
job: completed 35 2024-02-09 01:09:34.88717 +0530 IST m=+10.008558584
job: completed 35 2024-02-09 01:09:34.887194 +0530 IST m=+10.008582834
job: completed 35 2024-02-09 01:09:34.887188 +0530 IST m=+10.008576293
job: completed 35 2024-02-09 01:09:34.88719 +0530 IST m=+10.008578709
job: completed 35 2024-02-09 01:09:34.887193 +0530 IST m=+10.008581793
job: completed 36 2024-02-09 01:09:34.887199 +0530 IST m=+10.008587959
job: completed 35 2024-02-09 01:09:34.887176 +0530 IST m=+10.008564293
job: completed 42 2024-02-09 01:09:36.888285 +0530 IST m=+12.009681876
job: completed 43 2024-02-09 01:09:36.888459 +0530 IST m=+12.009855126
job: completed 44 2024-02-09 01:09:36.888475 +0530 IST m=+12.009871501
job: completed 45 2024-02-09 01:09:36.888482 +0530 IST m=+12.009878501
job: completed 46 2024-02-09 01:09:36.88849 +0530 IST m=+12.009886793
job: completed 47 2024-02-09 01:09:36.888497 +0530 IST m=+12.009893584
job: completed 48 2024-02-09 01:09:36.888507 +0530 IST m=+12.009903126
job: completed 49 2024-02-09 01:09:38.889494 +0530 IST m=+14.010898834
job: completed 49 2024-02-09 01:09:38.88954 +0530 IST m=+14.010944293
job: completed 49 2024-02-09 01:09:38.889575 +0530 IST m=+14.010979168
job: completed 49 2024-02-09 01:09:38.889561 +0530 IST m=+14.010965751
job: completed 49 2024-02-09 01:09:38.889567 +0530 IST m=+14.010971293
job: completed 49 2024-02-09 01:09:38.889571 +0530 IST m=+14.010975251
job: completed 49 2024-02-09 01:09:38.889554 +0530 IST m=+14.010958126
job: completed 56 2024-02-09 01:09:40.890683 +0530 IST m=+16.012095376
job: completed 56 2024-02-09 01:09:40.890696 +0530 IST m=+16.012108418
job: completed 56 2024-02-09 01:09:40.890724 +0530 IST m=+16.012136209
job: completed 56 2024-02-09 01:09:40.890727 +0530 IST m=+16.012139501
job: completed 56 2024-02-09 01:09:40.890729 +0530 IST m=+16.012141168
job: completed 56 2024-02-09 01:09:40.890734 +0530 IST m=+16.012146209
job: completed 56 2024-02-09 01:09:40.890733 +0530 IST m=+16.012145001
job: completed 63 2024-02-09 01:09:42.891807 +0530 IST m=+18.013226668
job: completed 63 2024-02-09 01:09:42.891831 +0530 IST m=+18.013250793
job: completed 63 2024-02-09 01:09:42.891838 +0530 IST m=+18.013258126
job: completed 63 2024-02-09 01:09:42.891841 +0530 IST m=+18.013260793
job: completed 63 2024-02-09 01:09:42.891843 +0530 IST m=+18.013262793
job: completed 63 2024-02-09 01:09:42.891845 +0530 IST m=+18.013265209
job: completed 63 2024-02-09 01:09:42.891849 +0530 IST m=+18.013269251
job: completed 70 2024-02-09 01:09:44.893001 +0530 IST m=+20.014429418
job: completed 71 2024-02-09 01:09:44.893096 +0530 IST m=+20.014524293
job: completed 71 2024-02-09 01:09:44.893091 +0530 IST m=+20.014519168
job: completed 72 2024-02-09 01:09:44.893126 +0530 IST m=+20.014554084
job: completed 72 2024-02-09 01:09:44.893114 +0530 IST m=+20.014542293
job: completed 72 2024-02-09 01:09:44.893115 +0530 IST m=+20.014543418
job: completed 71 2024-02-09 01:09:44.8931 +0530 IST m=+20.014527793
job: completed 77 2024-02-09 01:09:46.894258 +0530 IST m=+22.015693793
job: completed 77 2024-02-09 01:09:46.894199 +0530 IST m=+22.015635376
job: completed 79 2024-02-09 01:09:46.894378 +0530 IST m=+22.015814334
job: completed 79 2024-02-09 01:09:46.894384 +0530 IST m=+22.015820084
job: completed 80 2024-02-09 01:09:46.894397 +0530 IST m=+22.015833126
job: completed 81 2024-02-09 01:09:46.894407 +0530 IST m=+22.015842626
job: completed 82 2024-02-09 01:09:46.894411 +0530 IST m=+22.015846876
job: completed 84 2024-02-09 01:09:48.894618 +0530 IST m=+24.016061543
job: completed 84 2024-02-09 01:09:48.89461 +0530 IST m=+24.016054334
job: completed 85 2024-02-09 01:09:48.894645 +0530 IST m=+24.016088418
job: completed 86 2024-02-09 01:09:48.894647 +0530 IST m=+24.016090584
job: completed 87 2024-02-09 01:09:48.894649 +0530 IST m=+24.016092334
job: completed 85 2024-02-09 01:09:48.894646 +0530 IST m=+24.016089418
job: completed 88 2024-02-09 01:09:48.894652 +0530 IST m=+24.016096168
job: completed 91 2024-02-09 01:09:50.895695 +0530 IST m=+26.017146418
job: completed 91 2024-02-09 01:09:50.895723 +0530 IST m=+26.017174459
job: completed 91 2024-02-09 01:09:50.895675 +0530 IST m=+26.017127251
job: completed 91 2024-02-09 01:09:50.895713 +0530 IST m=+26.017165168
job: completed 91 2024-02-09 01:09:50.895717 +0530 IST m=+26.017168334
job: completed 91 2024-02-09 01:09:50.895712 +0530 IST m=+26.017163834
job: completed 91 2024-02-09 01:09:50.895708 +0530 IST m=+26.017159709
job: completed 98 2024-02-09 01:09:52.896781 +0530 IST m=+28.018240501
job: completed 98 2024-02-09 01:09:52.896796 +0530 IST m=+28.018255709
job: completed 98 2024-02-09 01:09:52.896807 +0530 IST m=+28.018266959
job: completed 98 2024-02-09 01:09:52.896826 +0530 IST m=+28.018286084
job: completed 98 2024-02-09 01:09:52.896834 +0530 IST m=+28.018293626
job: completed 98 2024-02-09 01:09:52.896841 +0530 IST m=+28.018300376
job: completed 98 2024-02-09 01:09:52.896855 +0530 IST m=+28.018314751

**Comment by griimick** [comment]
https://go.dev/doc/faq#closures_and_goroutines

**Comment by rudraksh97** [comment]
Upgrading the Go version to 1.22.0 fixed this as suggested in the link provided by @griimick

**Comment by rudraksh97** [comment]
https://go.dev/doc/faq#closures_and_goroutines

What happens with closures running as goroutines?

Due to the way loop variables work, before Go version 1.22 (see the end of this section for an update), some confusion could arise when using closures with concurrency. Consider the following program:

```
func main() {
    done := make(chan bool)

    values := []string{"a", "b", "c"}
    for _, v := range values {
        go func() {
            fmt.Println(v)
            done <- true
        }()
    }

    // wait for all goroutines to complete before exiting
    for _ = range values {
        <-done
    }
}
```

One might mistakenly expect to see a, b, c as the output. What you'll probably see instead is c, c, c. This is because each iteration of the loop uses the same instance of the variable v, so each closure shares that single variable. When the closure runs, it prints the value of v at the time fmt.Println is executed, but v may have been modified since the goroutine was launched. To help detect this and other problems before they happen, run [go vet](https://go.dev/cmd/go/#hdr-Run_go_tool_vet_on_packages).

To bind the current value of v to each closure as it is launched, one must modify the inner loop to create a new variable each iteration. One way is to pass the variable as an argument to the closure:

```
    for _, v := range values {
        go func(u string) {
            fmt.Println(u)
            done <- true
        }(v)
    }
```
    
In this example, the value of v is passed as an argument to the anonymous function. That value is then accessible inside the function as the variable u.

Even easier is just to create a new variable, using a declaration style that may seem odd but works fine in Go:

```
    for _, v := range values {
        v := v // create a new 'v'.
        go func() {
            fmt.Println(v)
            done <- true
        }()
    }
```

This behavior of the language, not defining a new variable for each iteration, was considered a mistake in retrospect, and has been addressed in [Go 1.22](https://go.dev/wiki/LoopvarExperiment), which does indeed create a new variable for each iteration, eliminating this issue.
